src/renderers/shared/fiber/ReactChildFiber.js

```diff
@@ -173,14 +158,38 @@ function ChildReconciler(shouldClone) {
     }
 
     if (isArray(newChildren)) {
-      var first : ?Fiber = null;
-      var prev : ?Fiber = null;
-      var existing : ?Fiber = existingChild;
+      // Nested arrays have a special fiber type.
+      const fragment = createFiberFromFragment(newChildren, priority);
+      fragment.return = returnFiber;
+      return fragment;
+    } else {
+      // TODO: Throw for unknown children.
+      return null;
+    }
+  }
+
+  // TODO: This API won't work because we'll need to transfer the side-effects of
+  // unmounting children to the returnFiber.
+  function reconcileChildFibers(
+    returnFiber : Fiber,
+    currentFirstChild : ?Fiber,
+    newChild : ReactNodeList,
+    priority : PriorityLevel
+  ) : ?Fiber {
+    // This function is not recursive.
+    // If the top level item is an array, we treat it as a set of children,
+    // not as a fragment. Nested arrays on the other hand will be treated as
+    // fragment nodes. Recursion happens at the normal flow.
```

Comment:
This function doesn't recurse down nested arrays by itself. It yields to the outer work loop by returning a fragment for nested arrays.